rfcs/20200113-tf-data-service.md

```diff
@@ -0,0 +1,591 @@
+than others. The tf.data service provides a mechanism for dynamically sharding,
+reducing the data imbalance across accelerators.
+
+### Visitation guarantees
+
+Model accuracy can often be improved when each training sample is trained on
+exactly once per epoch. The tf.data service can coordinate across workers to
+provide this guarantee.
+
+## Design Proposal
+
+The tf.data service is a master-worker system which iterates through datasets,
+producing outputs to be consumed by accelerators. The service is comprised of a
+few components:
+
+*   User-facing Python API for interacting with the tf.data service.
+*   Dataset splitting API for determining how to split up datasets for parallel
+    processing.
+*   Master and worker gRPC services.
+
+### Architecture
+
+The tf.data service is comprised of master and worker gRPC services which could
+be run in a couple of different configurations:
+
+#### Glossary
+
+**Master**: The single master coordinating the tf.data service.
+
+**Worker**: A tf.data service worker which performs dataset processing and
+provides dataset elements to consumers over RPC.
+
+**Consumer**: A machine which consumes data from the tf.data service. The
+consumer may be attached to a GPU or TPU, or use data for on-CPU training.
+
+#### Separate Cluster Architecture
+
+Each server is run on a separate host from the TensorFlow cluster. This
+configuration gives users a way to provide horizontally scaling CPU for
+processing their input pipelines and quickly feeding data to accelerators.
+
+#### Embedded Cluster Architecture
+
+without needing to provision additional compute resources. and gives all the
+benefits of the tf.data service except for horizontal scaling.
+
+#### Hybrid Architecture
+
+Users could run tf.data workers embedded in their TensorFlow cluster, and also
+run additional tf.data workers (and potentially the tf.data master) outside the
+cluster. This allows for horizontal worker scaling, while still leveraging the
+compute resources of the TensorFlow cluster for input processing.
+
+### User-facing Python API
+
+This API is how users will interact with the tf.data service from their Python
+code.
+
+```python
+def tf.data.experimental.service.distribute(address):
+  """Marks that a dataset should be processed by the tf.data service.
+
+  ds = ... # dataset to distribute
+  ds = ds.apply(tf.data.experimental.service.distribute(address))
+
+  Args:
+    address: The address of the tf.data service master.
+
+  Returns:
+    A function that can be passed to `dataset.apply()`.
+  """
+
+def tf.data.experimental.service.create_iteration(
+    dataset, num_consumers=1, num_tasks=None, deterministic=False):
+  """Begins distributed iteration over a dataset.
+
+  It is expected that the dataset contains at least one `.distribute(address)`
+  transformation, otherwise this method will print a warning and do nothing.
+
+  `create_iteration` will first register the dataset with the tf.data service
```

Comment:
How does create_iteration find the service?

Reply:
The service address is configured within the dataset by calling `dataset.apply(tf.data.experimental.service.distribute(address))`

Reply:
We should consider allowing users to specify a ClusterResolver here

Reply:
Agreed. A ClusterResolver would be helpful.